README.md
# Angular Logger

> Lightweight and configurable Angular logger

[![Build Status](https://travis-ci.org/Angular-RU/angular-logger.svg?branch=master)](https://travis-ci.org/Angular-RU/angular-logger)
[![npm version](https://badge.fury.io/js/%40angular-ru%2Flogger.svg)](https://badge.fury.io/js/%40angular-ru%2Flogger)
[![Coverage Status](https://coveralls.io/repos/github/Angular-RU/angular-logger/badge.svg?branch=develop)](https://coveralls.io/github/Angular-RU/angular-logger?branch=develop)
[![npm-stat](https://img.shields.io/npm/dt/@angular-ru/logger.svg)](https://npm-stat.com/charts.html?package=@angular-ru/logger&from=2017-01-12)

```typescript
import { LoggerModule } from '@angular-ru/logger';
...

@NgModule({
 imports: [
    LoggerModule.forRoot()
 ],
 ...
})
export class AppModule {}
```

## Motivation

This logger is a handy tool that can be useful in the design and development of the enterprise application level. Easy
setting of logging levels and convenient work with groups. Among other things, you can use meta programming
(decorators).

## Table of contents

-   [Logging](#)
    -   [Basic usage API `trace`, `debug`, `info`, `warn`, `error`](#example-basic-methods)
    -   [Groups, `groupCollapsed`, `collapsible`](#example-groups)
    -   [Nested groups (usage pipe method)](#example-nested-groups)
    -   [Set logging level (worked in single or groups)](#example-set-minimal-logging-level)
    -   [Customization style line](#example-set-style-line)
    -   [Customization global style line](#example-set-global-style-line)
    -   [Add css classes](#example-css-classes)
    -   [Output pretty json `stringify`](#example-pretty-json)
    -   [Copy `json, object, text` to clipboard](#example-clipboard)
    -   [Basic decorators](#example-decorators)
    -   [Decorator groups](#example-decorator-groups)
    -   [Decorator groups with function title](#example-decorator-group-with-function-title)
    -   [Configuration `Angular Logger`](#example-full-configurations)

*   [Todo](#todo)

## Logging

```
$ npm install @angular-ru/logger --save
```

```typescript
import { LoggerModule } from '@angular-ru/logger';
...

@NgModule({
 imports: [
    LoggerModule.forRoot()
 ],
 ...
})
export class AppModule {}
```

**Online examples**: https://stackblitz.com/github/Angular-RU/ng-logger

![](https://habrastorage.org/webt/lq/a9/_s/lqa9_sp8gxkwax_sy6x9w3qf5ry.gif)

### Example: basic methods

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.trace('trace is worked', 1, { a: 1 });
        this.logger.debug('debug is worked', 2, {});
        this.logger.info('info is worked', 3, Object);
        this.logger.warn('warn is worked', 4, String);
        this.logger.error('error is worked', 5, (2.55).toFixed());
    }
}
```

-   **Default level: All**

![](https://habrastorage.org/webt/0u/yj/1t/0uyj1tli-mzh0cor1cg4jwphsdk.png)

-   **Disable trace on console (filter):**

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    private readonly traceIsWork: string = 'trace is worked';
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.group(
            'Show trace in opened group',
            ({ trace }: LoggerService): void => {
                for (let i: number = 0; i < 20; i++) {
                    trace(this.traceIsWork, i);
                }
            }
        );
    }
}
```

### Example: groups

-   **Logger groups with auto closed (usage callback):**

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.groupCollapsed('EXAMPLE 2: show stack', () => {
            this.logger.trace(this.traceIsWork, 1, { a: 1 });
            this.logger.debug(this.debugIsWork, 2, console);
            this.logger.info(this.infoIsWork, 3, Object);
            this.logger.warn(this.warnIsWork, 4, String);
            this.logger.error(this.errorIsWork, 5, (2.55).toFixed());
        });

        this.logger.group(
            'Show trace in opened group',
            ({ trace }: LoggerService): void => {
                for (let i: number = 0; i < 20; i++) {
                    trace(this.traceIsWork, i);
                }
            }
        );

        this.logger.groupCollapsed(
            'Show trace in collapsed group',
            ({ debug }: LoggerService): void => {
                for (let i: number = 0; i < 15; i++) {
                    debug(this.traceIsWork, i);
                }
            }
        );
    }
}
```

![](https://habrastorage.org/webt/oo/ob/uh/ooobuhwfa3ncpctwirtirgmth6y.png)

### Example: nested groups

-   **Logger nested groups (with pipe):**

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger
            .groupCollapsed('GROUP TEST')
            .pipe(({ trace, debug, info, warn, error }: LoggerService) => {
                trace(this.traceIsWork);
                debug(this.debugIsWork);
                info(this.infoIsWork);
                warn(this.warnIsWork);
                error(this.errorIsWork);
            })
            .close();

        this.logger
            .group('A')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .groupCollapsed('B')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .group('C')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .closeAll();
    }
}
```

![](https://habrastorage.org/webt/_n/wz/8l/_nwz8l25o12tmu0d8sxyyjttmck.gif)

### Example: set minimal logging level

Basic parameterization

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.trace(this.traceIsWork, 1, { a: 1 });
        this.logger.debug(this.debugIsWork, 2, console);
        this.logger.info(this.infoIsWork, 3, Object);
        this.logger.warn(this.warnIsWork, 4, String);
        this.logger.error(this.errorIsWork, 5, (2.55).toFixed());

        this.logger.level = LoggerLevel.INFO;
        this.logger.log('Set new logger level');

        this.logger.trace(this.traceIsWork, 1, { a: 1 });
        this.logger.debug(this.debugIsWork, 2, console);
        this.logger.info(this.infoIsWork, 3, Object);
        this.logger.warn(this.warnIsWork, 4, String);
        this.logger.error(this.errorIsWork, 5, (2.55).toFixed());
    }
}
```

![](https://habrastorage.org/webt/0r/ya/xn/0ryaxnmaedlbc14imvodsezq4lg.png)

-   **Logger level groups (pretty usage API):**

```typescript
import { LoggerService, LoggerLevel } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.level = LoggerLevel.INFO;

        this.logger.trace
            .group('A')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .debug.group('B')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .info.group('C')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .warn.group('D')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .error.group('E')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close();

        this.logger.level = LoggerLevel.ALL;
    }
}
```

![](https://habrastorage.org/webt/x-/lm/pl/x-lmplcexk_nd0icuqe6ehslub4.png)

### Example: set style line

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.clear();

        this.logger.css('text-transform: uppercase; font-weight: bold').debug('window current ', window);
        this.logger.css('color: red; text-decoration: underline; font-weight: bold').info('It is awesome logger');
        this.logger.debug({ a: 1 });

        this.logger.warn(setStyle);
        this.logger.info('For global configuration, use the constructor parameters');
    }
}
```

![](https://habrastorage.org/webt/8b/dn/nv/8bdnnvsgj1m2rxnssviqptks874.png)

### Example: set global style line

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.clear();

        this.logger.css('font-weight: normal; text-decoration: none; font-style: italic').info(3.14);
        this.logger.css('font-weight: normal;').info(3.14);
        this.logger.warn('global format with style!');
    }
}
```

![](https://habrastorage.org/webt/y4/wm/vz/y4wmvzvsmtzt6zdqjcupxqmvodm.png)

### Example: CSS classes

```typescript
import { LoggerModule } from '@angular-ru/logger';

@NgModule({
    // ..
    imports: [
        LoggerModule.forRoot({
            cssClassMap: {
                bold: 'font-weight: bold',
                'line-through': 'text-decoration: line-through',
                'code-sandbox': `
                  color: #666;
                  background: #f4f4f4;
                  border-left: 3px solid #f36d33;
                  font-family: monospace;
                  font-size: 15px;`
            }
        })
    ]
    // ..
})
```

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.logger.cssClass('bold line-through').log('JavaScript sucks', 'JavaScript is the best');

        this.logger
            .cssClass('code-sandbox')
            .log('\n   @Component({ .. })' + '\n   export class AppComponent { .. }    \n\n');

        this.logger.cssClass('bold line-through').debug('JavaScript sucks', 'JavaScript is the best');
    }
}
export class AppModule {}
```

![](https://habrastorage.org/webt/d5/tm/aa/d5tmaaomjql5px_wkzxnodhacnk.png)

### Example: pretty json

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        // default browser print json
        this.logger.debug('Classic output json', jsonExample);

        // for pretty json usage logger.log method
        this.logger.log(...this.logger.prettyJSON(jsonExample));
    }
}
```

![](https://habrastorage.org/webt/eo/ej/k5/eoejk5t_hqvo2xeaitkbzm43grm.png)

### Example: clipboard

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {
    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        const example: string = 'test string';
        this.logger.log(example);
        this.logger.copy(example);
    }
}
```

![](https://habrastorage.org/webt/sd/uo/3a/sduo3ags-5rsnoanvxtckysresw.gif)

### Example: decorators

```typescript
import { LoggerService, Logger, DebugLog, TraceLog, InfoLog, WarnLog, ErrorLog, Log, LogFn } from '@angular-ru/logger';

export class AppComponent {
    @Logger() public logger: LoggerService;
    @TraceLog() public trace: LogFn;
    @DebugLog() public debug: LogFn;
    @InfoLog() public info: LogFn;
    @ErrorLog() public error: LogFn;
    @WarnLog() public warn: LogFn;
    @Log() public log: LogFn;

    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    public showExample(): void {
        this.logger.clear();
        this.logger.log('log is worked');
        this.trace(this.traceIsWork, 1, { a: 1 });
        this.debug(this.debugIsWork, 2, console);
        this.info(this.infoIsWork, 3, Object);
        this.warn(this.warnIsWork, 4, String);
        this.error(this.errorIsWork, 5, (2.55).toFixed());
    }
}
```

![](https://habrastorage.org/webt/fk/ar/a5/fkara5xhh75iz1q9_dales4cu9s.png)

### Example: decorator groups

```typescript
import { LoggerService, Logger, LoggerLevel, Group } from '@angular-ru/logger';

export class AppComponent {
    @Logger() public logger: LoggerService;

    @Group('test title', LoggerLevel.WARN)
    private helloWorld(name: string): string {
        this.logger.log('log only in group', name);
        return 'hello world';
    }

    public showExample11(): void {
        this.logger.log(this.helloWorld('Hello'));
    }
}
```

![](https://habrastorage.org/webt/na/bu/zw/nabuzwdhsanhy0sznh5tvwdu8es.png)

### Example: decorator group with function title

```typescript
import { Log, LogFn, Group } from '@angular-ru/logger';

export class AppComponent {
    @Log() public log: LogFn;

    @Group((name: string) => `Test group with ${name}`)
    public method(name: string): string {
        this.log('group is worked');
        return name;
    }

    public showExample(): void {
        this.method('hello world');
    }
}
```

![](https://habrastorage.org/webt/j9/mz/4v/j9mz4vbyhi0dg_8kr5wjmfwgiye.png)

### Example: timer decorator

```typescript
import { Log, LogFn, TimerLog, LoggerLevel, LoggerService, Logger } from '@angular-ru/logger';
export class AppComponent {
    @Log() public log: LogFn;
    @Logger() public logger: LoggerService;

    @TimerLog('Test timer')
    public showExample(): void {
        this.logger.clear();
        this.log('test log');
    }

    @TimerLog('Advanced timer', LoggerLevel.WARN, false)
    public showExample(): void {
        this.logger.clear();
        this.log('Advanced test log');
    }
}
```

![](https://habrastorage.org/webt/ur/zl/66/urzl66me9nixsnauhsfpge2tn0a.png)

### Example: full configurations

```typescript
import { LoggerModule } from '@angular-ru/logger';

@NgModule({
    // ..
    imports: [
        LoggerModule.forRoot({
            useLevelGroup: true,
            globalLineStyle: 'color: red; text-decoration: underline; font-weight: bold; font-size: 15px',
            cssClassMap: {
                bold: 'font-weight: bold',
                'line-through': 'text-decoration: line-through',
                'code-sandbox': `
                  color: #666;
                  background: #f4f4f4;
                  border-left: 3px solid #f36d33;
                  font-family: monospace;
                  font-size: 15px;`
            },
            labelNames: {
                [LoggerLevel.TRACE]: '[trace]',
                [LoggerLevel.DEBUG]: '[debug]',
                [LoggerLevel.INFO]: '[info]',
                [LoggerLevel.WARN]: '[warn]',
                [LoggerLevel.ERROR]: '[error]'
            },
            labelColors: {
                [LoggerLevel.TRACE]: 'violet',
                [LoggerLevel.DEBUG]: 'black',
                [LoggerLevel.INFO]: 'tomato',
                [LoggerLevel.WARN]: 'green',
                [LoggerLevel.ERROR]: 'cyan'
            }
        })
    ]
    // ..
})
export class AppModule {}
```

```typescript
import { LoggerService } from '@angular-ru/logger';

export class AppComponent implements OnInit {

  public ngOnInit(): void {
    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    constructor(private readonly logger: LoggerService) {}

    public showExample(): void {
        this.logger.log('Example');
        this.logger.trace(this.traceIsWork, 1, { a: 1 });
        this.logger.debug(this.debugIsWork, 2, console);
        this.logger.info(this.infoIsWork, 3, Object);
        this.logger.warn(this.warnIsWork, 4, String);
        this.logger.error(this.errorIsWork, 5, (2.55).toFixed());
    }
}
```

![](https://habrastorage.org/webt/we/vi/xw/wevixwtlrik5gqxfif2wadprura.png)

## Todo

-   [x] Override console
-   [x] Logger method (trace, debug, info, warning, error)
-   [x] Logger group + groupCollapsible (pipes)
-   [x] Logger pretty write object
-   [x] Set style by css
-   [x] Logger level groups (trace, debug, info, warn, error)
-   [x] Clipboard data
-   [x] Set global style
-   [x] Added css classes
-   [x] Dependency Injection for Angular
-   [x] Switch enable/disable default console output
-   [x] Decorators
-   [x] Timers (decorator)
-   [ ] Profiling (memory usage, sizeof, time execute)
-   [ ] Pre process output
-   [ ] Format output console

## Authors

[Eleonora Zbarskaya](https://github.com/kingofferelden), [Ivanov Maxim](https://github.com/splincode)


CHANGELOG.md
#### 0.0.0 (2019-05-31)

##### Build System / Dependencies

* **deps:**  bump tar from 2.2.1 to 2.2.2 (#5) (95fa24d7)
*  coverage pipeline (ce31cba4)
*  add realise tools (a2bd14d6)

##### Chores

*  small fix in path (bccfbce3)
*  fix by prettier (938e4cbd)
*  update package json (c7d5e95b)
*  update package json (4eb14b25)
*  move helpers, change loops (62dd41a3)
*  change loops (709cdc19)
*  fix type (41af75ea)
*  remove jest config from package.json (010c1118)
*  format files (e885c0b2)
*  move type (4a55b718)
*  update .gitignore (d402b98f)
*  fix packages (91eb0102)

##### Continuous Integration

*  bump node version (5e835d0c)

##### Documentation Changes

*  change navigation (4347278e)
*  change example (9a04cc8b)
*  change example (0f2d0f2b)
*  add examples to readme and preview (54955e91)
*  update README.md (3192d892)
*  small fix in readme (7eabb6c0)
*  small fix in readme (5cb735e0)
*  small fix in readme (1779202a)
*  add examples to Readme (8eaa3640)

##### New Features

*  add timer decorator (e219a138)
*  add husky (6c05100b)
*  add global configs (2c872793)
*  add css classes and global style lines (66c4aaa8)
*  add level groups (db4836d3)
*  add level groups (d86a7d11)
*  add json factory, clipboard factory (8825ce5d)
*  add clipboard service (5cea592e)
*  add json-factory service (8f6c90ce)
*  add jest (3e5a9185)
*  add prettier (f9563462)
*  add custom css (7d5d120f)

##### Other Changes

*  add decorators (579bb843)

##### Refactors

*  fix naming after codereview (39c3e6dc)
*  fix const (87e76195)
*  fix after code review (d57a0c73)
*  fix sonar issues (725f35b9)
*  fix type after code review (659edc0f)
*  small fix (5ad72a7d)
*  change path in timer-factory (311bfcc0)
*  change after code review (599af2fb)
*  change after codereview (3e20b8ba)
*  small fix after code review (e6edc773)
*  small fix after code review (aacb30b8)
*  small fix after sonar (aedd6cdb)
*  small fixes after sonar (eb0cc8ac)
*  small fixes (92944d60)
*  small fixes after reviews (574a50d4)
*  small fixes after reviews (4d85934e)
*  small fixes after reviews (971fa072)

##### Tests

*  add tests for timer decorator (133d410b)
*  add tests for decorators (b2a2fc06)
*  remove duplicate literals in css-parcer (12cc068f)
*  add tests (f6b05cf7)
*  add clipboard, css, group unit tests (0d24fbec)

#### 0.0.0 (2019-05-31)

##### Build System / Dependencies

* **deps:**  bump tar from 2.2.1 to 2.2.2 (#5) (95fa24d7)
*  coverage pipeline (ce31cba4)
*  add realise tools (a2bd14d6)

##### Chores

*  small fix in path (bccfbce3)
*  fix by prettier (938e4cbd)
*  update package json (c7d5e95b)
*  update package json (4eb14b25)
*  move helpers, change loops (62dd41a3)
*  change loops (709cdc19)
*  fix type (41af75ea)
*  remove jest config from package.json (010c1118)
*  format files (e885c0b2)
*  move type (4a55b718)
*  update .gitignore (d402b98f)
*  fix packages (91eb0102)

##### Continuous Integration

*  bump node version (5e835d0c)

##### Documentation Changes

*  change navigation (4347278e)
*  change example (9a04cc8b)
*  change example (0f2d0f2b)
*  add examples to readme and preview (54955e91)
*  update README.md (3192d892)
*  small fix in readme (7eabb6c0)
*  small fix in readme (5cb735e0)
*  small fix in readme (1779202a)
*  add examples to Readme (8eaa3640)

##### New Features

*  add timer decorator (e219a138)
*  add husky (6c05100b)
*  add global configs (2c872793)
*  add css classes and global style lines (66c4aaa8)
*  add level groups (db4836d3)
*  add level groups (d86a7d11)
*  add json factory, clipboard factory (8825ce5d)
*  add clipboard service (5cea592e)
*  add json-factory service (8f6c90ce)
*  add jest (3e5a9185)
*  add prettier (f9563462)
*  add custom css (7d5d120f)

##### Other Changes

*  add decorators (579bb843)

##### Refactors

*  fix naming after codereview (39c3e6dc)
*  fix const (87e76195)
*  fix after code review (d57a0c73)
*  fix sonar issues (725f35b9)
*  fix type after code review (659edc0f)
*  small fix (5ad72a7d)
*  change path in timer-factory (311bfcc0)
*  change after code review (599af2fb)
*  change after codereview (3e20b8ba)
*  small fix after code review (e6edc773)
*  small fix after code review (aacb30b8)
*  small fix after sonar (aedd6cdb)
*  small fixes after sonar (eb0cc8ac)
*  small fixes (92944d60)
*  small fixes after reviews (574a50d4)
*  small fixes after reviews (4d85934e)
*  small fixes after reviews (971fa072)

##### Tests

*  add tests for timer decorator (133d410b)
*  add tests for decorators (b2a2fc06)
*  remove duplicate literals in css-parcer (12cc068f)
*  add tests (f6b05cf7)
*  add clipboard, css, group unit tests (0d24fbec)



src/app/app.component.ts
/* tslint:disable:no-duplicate-string */
import { Component, OnInit, ViewEncapsulation } from '@angular/core';
import {
    DebugLog,
    ErrorLog,
    Group,
    InfoLog,
    LogFn,
    Logger,
    LoggerLevel,
    Log,
    LoggerService,
    TimerLog,
    TraceLog,
    WarnLog
} from '@angular-ru/logger';
import * as devtools from 'devtools-detect';

@Component({
    selector: 'app-root',
    templateUrl: './app.component.html',
    styleUrls: ['./app.component.css'],
    encapsulation: ViewEncapsulation.None
})
export class AppComponent implements OnInit {
    @Logger() public loggerInjection: LoggerService;
    @TraceLog() public trace: LogFn;
    @DebugLog() public debug: LogFn;
    @InfoLog() public info: LogFn;
    @ErrorLog() public error: LogFn;
    @WarnLog() public warn: LogFn;
    @Log() public log: LogFn;

    public isLoaded: boolean;
    public devToolsIsOpen: boolean = devtools.isOpen;

    private readonly traceIsWork: string = 'trace is worked';
    private readonly debugIsWork: string = 'debug is worked';
    private readonly infoIsWork: string = 'info is worked';
    private readonly warnIsWork: string = 'warn is worked';
    private readonly errorIsWork: string = 'error is worked';

    constructor(private readonly logger: LoggerService) {}

    public ngOnInit(): void {
        this.isLoaded = true;
        window.addEventListener('devtoolschange', (e: devtools.DevToolsEvent) => {
            this.devToolsIsOpen = e.detail.isOpen;
        });
    }

    public showExample1(): void {
        this.logger.clear();
        this.log('log is worked');
        this.trace(this.traceIsWork, 1, { a: 1 });
        this.debug(this.debugIsWork, 2, console);
        this.info(this.infoIsWork, 3, Object);
        this.warn(this.warnIsWork, 4, String);
        this.error(this.errorIsWork, 5, (2.55).toFixed());
    }

    public showExample2(): void {
        this.logger.clear();

        this.logger.groupCollapsed('EXAMPLE 2: show stack', () => {
            this.trace(this.traceIsWork, 1, { a: 1 });
            this.debug(this.debugIsWork, 2, console);
            this.info(this.infoIsWork, 3, Object);
            this.warn(this.warnIsWork, 4, String);
            this.error(this.errorIsWork, 5, (2.55).toFixed());
        });

        this.logger.group(
            'Show trace in opened group',
            ({ trace }: LoggerService): void => {
                for (let i: number = 0; i < 20; i++) {
                    trace(this.traceIsWork, i);
                }
            }
        );

        this.logger
            .groupCollapsed(
                'Show trace in collapsed group',
                ({ debug }: LoggerService): void => {
                    for (let i: number = 0; i < 15; i++) {
                        debug(this.traceIsWork, i);
                    }
                }
            )
            .closeAll();
    }

    public showExample3(): void {
        this.logger.clear();

        this.logger
            .groupCollapsed('GROUP TEST')
            .pipe(({ trace, debug, info, warn, error }: LoggerService) => {
                trace(this.traceIsWork);
                debug(this.debugIsWork);
                info(this.infoIsWork);
                warn(this.warnIsWork);
                error(this.errorIsWork);
            })
            .close();

        this.logger
            .group('A')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .groupCollapsed('B')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .group('C')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .closeAll();
    }

    public showExample4(): void {
        this.logger.clear();

        this.logger.level = LoggerLevel.INFO;

        this.logger.log('log is working', 1, String);
        this.trace(this.traceIsWork, 4, String);
        this.debug(this.debugIsWork, 4, String);
        this.warn(this.warnIsWork, 4, String);
        this.error(this.errorIsWork, 5, (2.55).toFixed());

        this.logger.level = LoggerLevel.ALL;
    }

    public showExample5(): void {
        this.logger.clear();

        this.logger.css('text-transform: uppercase; font-weight: bold').debug('window current ', window);

        this.logger.css('color: red; text-decoration: underline; font-weight: bold').info('It is awesome logger');
        this.debug({ a: 1 });

        this.warn('logger.css(...) does not define a global format!');
        this.info('For global configuration, use the constructor parameters');
    }

    public showExample6(): void {
        this.logger.clear();

        const jsonExample: object = {
            id: 1,
            hello: 'world'
        };

        this.debug('Classic output json', jsonExample);

        this.logger.log(...this.logger.prettyJSON(jsonExample));
    }

    public showExample7(): void {
        this.logger.clear();

        const example: string = 'test string';

        this.logger.log(example);
        this.logger.copy(example);
    }

    public showExample8(): void {
        this.logger.clear();
        this.logger.level = LoggerLevel.INFO;

        this.trace
            .group('A')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .debug.group('B')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .info.group('C')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .warn.group('D')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close()

            .error.group('E')
            .pipe(
                ({ trace }: LoggerService) => trace(this.traceIsWork),
                ({ debug }: LoggerService) => debug(this.debugIsWork),
                ({ info }: LoggerService) => info(this.infoIsWork),
                ({ warn }: LoggerService) => warn(this.warnIsWork),
                ({ error }: LoggerService) => error(this.errorIsWork)
            )
            .close();

        this.logger.level = LoggerLevel.ALL;
    }

    public showExample9(): void {
        this.logger.clear();

        this.logger.css('font-weight: normal; text-decoration: none; font-style: italic;').info(3.14);
        this.logger.css('font-weight: normal;').info(3.14);
        this.warn('global format with style!');
    }

    public showExample10(): void {
        this.logger.clear();

        this.logger.cssClass('bold line-through').log('JavaScript sucks', 'JavaScript is the best');

        this.logger
            .cssClass('code-sandbox')
            .log('\n   @Component({ .. })' + '\n   export class AppComponent { .. }    \n\n');

        this.logger.cssClass('bold line-through').debug('JavaScript sucks', 'JavaScript is the best');

        this.logger.level = LoggerLevel.INFO;
    }

    public showExample11(): void {
        this.loggerInjection.clear();
        this.logger.log(this.helloWorld('Max'));
    }

    @Group((name: string) => `Test group with ${name}`)
    public method(name: string): string {
        this.loggerInjection.log('group is worked');
        return name;
    }

    public showExample12(): void {
        this.loggerInjection.clear();
        this.method('hello world');
    }

    @TimerLog('Test timer')
    public showExample13(): void {
        this.logger.clear();
        this.log('test log');
    }

    @TimerLog('Advanced timer', LoggerLevel.WARN, false)
    public showExample14(): void {
        this.logger.clear();
        this.log('Advanced test log');
    }

    @Group('test title', LoggerLevel.WARN)
    private helloWorld(name: string): string {
        this.logger.log('log only in group', name);
        return 'hello world';
    }
}
